library/declaremods.ml

```diff
@@ -708,14 +730,21 @@ let declare_modtype interp_modast id args mtys (mty,ann) fs =
   (* We simulate the beginning of an interactive module,
      then we adds the module parameters to the global env. *)
   let mp = Global.start_modtype id in
-  let arg_entries_r = intern_args interp_modast args in
+  let arg_entries_r, cst = intern_args interp_modast args in
+  let () = Global.push_context_set true cst in
   let params = mk_params_entry arg_entries_r in
   let env = Global.env () in
-  let mte, _ = interp_modast env ModType mty in
+  let mte, _, cst = interp_modast env ModType mty in
+  let () = Global.push_context_set true cst in
+  let env = Global.env () in
```

Comment:
These repeated push_context/get env feel kinda strange.

Reply:
Yes, this code is pretty ugly. But it's just making explicit that we should reimplement it using a purely functional API instead of randomly calling effectful functions. I'll leave that for another time though.